moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1169 @@
+          else -> throw syntaxError("Unexpected value")
+        }
+      }
+      // In lenient mode, a 0-length literal in an array means 'null'.
+      ';', ',' -> return when (peekStack) {
+        JsonScope.EMPTY_ARRAY, JsonScope.NONEMPTY_ARRAY -> {
+          checkLenient()
+          PEEKED_NULL.also { peeked = it }
+        }
+        else -> throw syntaxError("Unexpected value")
+      }
+      '\'' -> {
+        checkLenient()
+        buffer.readByte() // Consume '\''.
+        return PEEKED_SINGLE_QUOTED.also { peeked = it }
+      }
+      '"' -> {
+        buffer.readByte() // Consume '\"'.
+        return PEEKED_DOUBLE_QUOTED.also { peeked = it }
+      }
+      '[' -> {
+        buffer.readByte() // Consume '['.
+        return PEEKED_BEGIN_ARRAY.also { peeked = it }
+      }
+      '{' -> {
+        buffer.readByte() // Consume '{'.
+        return PEEKED_BEGIN_OBJECT.also { peeked = it }
+      }
+      else -> Unit /* no-op */
+    }
+    var result = peekKeyword()
+    if (result != PEEKED_NONE) {
+      return result
+    }
+    result = peekNumber()
+    if (result != PEEKED_NONE) {
+      return result
+    }
+    if (!isLiteral(buffer[0].toInt())) {
+      throw syntaxError("Expected value")
+    }
+    checkLenient()
+    return PEEKED_UNQUOTED.also { peeked = it }
+  }
+
+  @Throws(IOException::class)
+  private fun peekKeyword(): Int {
+    // Figure out which keyword we're matching against by its first character.
+    var c = buffer[0].asChar()
+    val keyword: String
+    val keywordUpper: String
+    val peeking: Int
+    when (c) {
+      't', 'T' -> {
+        keyword = "true"
+        keywordUpper = "TRUE"
+        peeking = PEEKED_TRUE
+      }
+      'f', 'F' -> {
+        keyword = "false"
+        keywordUpper = "FALSE"
+        peeking = PEEKED_FALSE
+      }
+      'n', 'N' -> {
+        keyword = "null"
+        keywordUpper = "NULL"
+        peeking = PEEKED_NULL
+      }
+      else -> return PEEKED_NONE
+    }
+
+    // Confirm that chars [1..length) match the keyword.
+    val length = keyword.length
+    for (i in 1 until length) {
+      val i_long = i.toLong()
+      if (!source.request((i_long + 1))) {
+        return PEEKED_NONE
+      }
+      c = buffer[i_long].asChar()
+      if (c != keyword[i] && c != keywordUpper[i]) {
+        return PEEKED_NONE
+      }
+    }
+    if (source.request((length + 1).toLong()) && isLiteral(buffer[length.toLong()].toInt())) {
+      return PEEKED_NONE // Don't match trues, falsey or nullsoft!
+    }
+
+    // We've found the keyword followed either by EOF or by a non-literal character.
+    buffer.skip(length.toLong())
+    return peeking.also { peeked = it }
+  }
+
+  @Throws(IOException::class)
+  private fun peekNumber(): Int {
+    var value: Long = 0 // Negative to accommodate Long.MIN_VALUE more easily.
+    var negative = false
+    var fitsInLong = true
+    var last = NUMBER_CHAR_NONE
+    var i: Long = 0
+    charactersOfNumber@ while (true) {
```

Comment:
I'm not 100% certain that the label is needed, but I've kept it to maintain as much parity w/ the Java sources as possible

Reply:
Doesn't seem needed

Reply:
fixed in 730b6f7